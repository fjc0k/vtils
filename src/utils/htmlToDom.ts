// https://github.com/fregante/doma/blob/master/index.ts

/**
 * 将 HTML 字符串转为 DocumentFragment。
 *
 * @param html HTML 字符串
 */
export function htmlToDocumentFragment(html: string): DocumentFragment {
  if (html == null) {
    return new DocumentFragment()
  }

  const template = document.createElement('template')
  template.innerHTML = html
  return template.content
}

/**
 * 将 HTML 字符串转为 Element。
 *
 * @param html HTML 字符串
 */
export function htmlToElement<T extends Element = Element>(
  html: string,
): T | undefined {
  return (htmlToDocumentFragment(html).firstElementChild as T) ?? undefined
}
