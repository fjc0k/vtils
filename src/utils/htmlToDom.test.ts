import { htmlToDocumentFragment, htmlToElement } from './htmlToDom'

describe('htmlToDom', () => {
  test('htmlToDocumentFragment 正常', () => {
    expect(
      // @ts-expect-error
      htmlToDocumentFragment(),
    ).toMatchSnapshot()
    expect(htmlToDocumentFragment('test')).toMatchSnapshot()
    expect(htmlToDocumentFragment('test <a>hello</a>')).toMatchSnapshot()
    expect(htmlToDocumentFragment('<a>hello</a>')).toMatchSnapshot()
    expect(
      htmlToDocumentFragment('<a>hello</a>')
        .querySelector('a')
        ?.textContent?.trim(),
    ).toMatchSnapshot()
  })

  test('htmlToElement 正常', () => {
    expect(htmlToElement('test')).toMatchSnapshot()
    expect(htmlToElement('test <a>hello</a>')).toMatchSnapshot()
    expect(htmlToElement('<a>hello</a>')).toMatchSnapshot()
  })
})
