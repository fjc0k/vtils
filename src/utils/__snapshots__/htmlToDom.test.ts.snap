// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`htmlToDom htmlToDocumentFragment 正常 1`] = `
<DocumentFragment>
  test
</DocumentFragment>
`;

exports[`htmlToDom htmlToDocumentFragment 正常 2`] = `
<DocumentFragment>
  test 
  <a>
    hello
  </a>
</DocumentFragment>
`;

exports[`htmlToDom htmlToDocumentFragment 正常 3`] = `
<DocumentFragment>
  <a>
    hello
  </a>
</DocumentFragment>
`;

exports[`htmlToDom htmlToDocumentFragment 正常 4`] = `"hello"`;

exports[`htmlToDom htmlToElement 正常 1`] = `undefined`;

exports[`htmlToDom htmlToElement 正常 2`] = `
<a>
  hello
</a>
`;

exports[`htmlToDom htmlToElement 正常 3`] = `
<a>
  hello
</a>
`;
